update rsk-starter-box (#351)
rsksmart/devportal: tutorials/truffle-boxes/rsk-starter-box.md

---
layout: rsk
title: Using Truffle box rsk-starter-box
tags: tutorial, rsk, truffle, truffle-box 
description: "How to use the Truffle box rsk-starter-box, which comes with everything you need to start using Truffle on RSK networks. It includes network configurations for Mainnet, Testnet and the SimpleStorage contract as an example to deploy."
---

In this tutorial I will show you step-by-step how to use the Truffle box [rsk-starter-box](https://github.com/rsksmart/rsk-starter-box), 
which comes with everything you need to start using Truffle on RSK networks. 
It includes network configurations for Mainnet, Testnet and the SimpleStorage contract as an example to deploy.

# Overview

Here is a summary of the steps we will take in this tutorial:

1. Setup prerequisites;
2. Install RSK Truffle Starter Box;
3. Learn how to use the Truffle development console;
4. Create a wallet;
5. Configure Truffle to connect to RSK network;
6. Get R-BTC;
7. Deploy a smart contract on RSK network using Truffle;
8. Interact with the smart contract at Truffle console.

> If you came from [Truffle box page rsk-starter-box](https://github.com/rsksmart/rsk-starter-box) 

Author: It is not only used Truffle-rsk-starter-box before, I would like to say that the person was on page `https://github.com/rsksmart/rsk-starter-box` and he ended up here because he clicked on a link that was on that page.
What is the better way to say this?

> If you came from [Truffle box page rsk-starter-box](https://github.com/rsksmart/rsk-starter-box) 

Reviewer: If you were redirected from the [Truffle-rsk-starter-box](https://github.com/rsksmart/rsk-starter-box) page, this tutorial provides more explanatory steps with examples for you to follow.

> If you came from [Truffle box page rsk-starter-box](https://github.com/rsksmart/rsk-starter-box) 

Author: Thanks!

* Code editor: Visual Studio Code (VSCode) or any other editor of your choice
* Truffle

## POSIX compliant shell

The **Portable Operating System Interface (POSIX)** is a family of standards specified by the IEEE Computer Society for maintaining compatibility between operating systems. POSIX defines the application programming interface (API), along with command line shells and utility interfaces, for software compatibility with variants of Unix and other operating systems.
Source: [Wikipedia](https://en.wikipedia.org/wiki/POSIX)

* Mac OSX and Linux distributions: Use the standard terminal
* Windows: If you use the standard `cmd` terminal, or PowerShell, the commands here may not work.
  Consider installing [Git for Windows](https://gitforwindows.org/), which comes with Git Bash bundled.
  Here is a [tutorial on installing and using Git Bash](https://www.atlassian.com/git/tutorials/git-bash).

## cURL

This is a system command that is likely already installed on your system,
which allows you to make network requests, such as HTTP requests,
from your command line.

If `curl --version` displays an error,
[download curl](https://curl.haxx.se/download.html).

> Tak a look in Windows SO:

Author: Sorry, it is OS!

(5) 449654cde095f2349113ef12a93e139b4302bc95adb3619d08adf53dde9b8847
(6) c217f12a89c352fc70b5f1bd5742314b4fb1bb1e35cb779fdb3c2390106355db
(7) 1d4c74dfa4e99e161130c18cc63938bb120a128cefbf1b9188efc678bf5722cb
(8) 0f44e0becf2e090db498a1b747d2a758fcc81fb0241f350d61117a9c6b1fa82e
(9) 85218c5eec657470dafeb09e6f7101f91d21bfe822fbeeecfc9275f798662a63

Mnemonic: virtual valve razor retreat either turn possible student grief engage attract fiber

⚠️  Important ⚠️  : This mnemonic was created for you by Truffle. It is not secure.
Ensure you do not use it on production blockchains, or else you risk losing funds.

truffle(develop)>  
```

> Inside the development console we don't preface commands with `truffle`.

# Compile a smart contract

In the Truffle console, run this command:

```javascript
compile
```

```windows-command-prompt
truffle(develop)> compile

Compiling your contracts...
===========================
> Compiling .\contracts\Migrations.sol
> Compiling .\contracts\SimpleStorage.sol
> Artifacts written to C:\RSK\rsk-starter\build\contracts
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang

truffle(develop)>
```

# Deploy a smart contract

> To deploy a smart contract using Truffle, we need a new migrations file where Truffle will find it, containing instructions to deploy the smart contract. 

Reviewer: ```suggestion
To deploy a smart contract using Truffle, we need a new migrations file where Truffle will find it, this file contains instructions to deploy the smart contract. 
```

The `migrations` folder has JavaScript files that help you deploy contracts to the network. 
These files are responsible for staging your deployment tasks, and they're written under the assumption that your deployment needs will change over time. 

> A history of previously run migrations is recorded on-chain through a special Migrations contract. 

Reviewer: ```suggestion
A history of previously ran migrations is recorded on-chain through a special Migrations contract. 
```

   > gas used:            225237 (0x36fd5)
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.00450474 ETH

   > Saving migration to chain.
   > Saving artifacts
   -------------------------------------
   > Total cost:          0.00450474 ETH

2_deploy_contracts.js
=====================

   Deploying 'SimpleStorage'
   -------------------------
   > transaction hash:    0x550a98af626b20d567422c1219ab6629aaaee742e4bfc81033386d9377700eb7
   > Blocks: 0            Seconds: 0
   > contract address:    0x73f2aa5D251DbdEd6C950257124eA93bb00c0Ec0
   > block number:        3
   > block timestamp:     1591365569
   > account:             0x1056F747cf4bC7710E178B2aeED4Eb8c8506c728
   > balance:             99.99272422
   > gas used:            96189 (0x177bd)
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.00192378 ETH

   > Saving migration to chain.
   > Saving artifacts
   -------------------------------------
   > Total cost:          0.00192378 ETH

Summary
=======
> Total deployments:   2
> Final cost:          0.00642852 ETH

truffle(develop)>  
```

# Test a smart contract

> Truffle has an automated testing framework to facilitate testing your contracts.

Reviewer: ```suggestion
Truffle has an automated testing framework to facilitate the testing of contracts.
```

All test files should be located in the `test` directory.
To learn more, go to the Truffle documentation, in the section [testing your contracts](https://www.trufflesuite.com/docs/truffle/testing/testing-your-contracts).

Our box also comes with the file `TestSimpleStorage.js` for testing the smart contract. 
You can check it out in the `test` folder.

```javascript
test
```

> ![test](/assets/img/tutorials/rsk-starter-box/image-01.png)

Reviewer 2: This is an image of text, please replace with text using the `windows-command-prompt`.
See examples from `_quick-start/step1-install-rsk-local-node.md` and `tutorials/ethereum-devs/setup-truffle-oz.md`.

Author: I would like to keep this content as an image, because of the colors. It is not a windows terminal, it is a truffle terminal.

```windows-command-prompt
truffle(develop)> test
Using network 'develop'.

Compiling your contracts...
===========================
> Compiling .\contracts\Migrations.sol
> Compiling .\contracts\SimpleStorage.sol
> Artifacts written to C:\Users\Solange\AppData\Local\Temp\test-202055-3640-nxpnxb.3exj
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang

  Contract: SimpleStorage
    √ should store a value (131ms)

  1 passing (164ms)

truffle(develop)>  
```

# Interact with a smart contract in development console

> Make sure you deploy the smart contract before executing this part.

The next commands will run inside the development console.

> ## Connect with our contract published

Reviewer: ```suggestion
## Connect with our published contract
```

```javascript
const simpleStorage = await SimpleStorage.deployed()
```

Now simpleStorage variable contains an instance of the previously deployed contract.

```windows-command-prompt
truffle(develop)> const simpleStorage = await SimpleStorage.deployed()
undefined
truffle(develop)>  
```

The published contract information is stored by default in the `build\contracts` folder. 
You will find a JSON file with the same name of our smart contract.
The section `networks` has the networks in which the smart contract was published, including its address and hash of the transaction.

![simpleStorage.json networks](/assets/img/tutorials/rsk-starter-box/image-02.png)

> The ABI - Application Binary Interface also is very important because to connect with one smart contract, we must know the ABI and the address of the smart contract.

Reviewer: ```suggestion
The ABI - Application Binary Interface is also very important because to connect with one smart contract, we must know the ABI and the address of the smart contract.
```

![simpleStorage.json abi](/assets/img/tutorials/rsk-starter-box/image-03.png)

## Get value

Get the value stored in the contract.

```javascript
simpleStorage.get().then(bn => bn.toNumber())
```

We do not have any value stored, because we do not define anything at the moment when we deployed.

```windows-command-prompt
truffle(develop)> simpleStorage.get().then(bn => bn.toNumber())
0
truffle(develop)>  
```

## Set value

Store some value in the contract.

```javascript
simpleStorage.set(10)
```

> ![simpleStorage set](/assets/img/tutorials/rsk-starter-box/image-04.png)

Reviewer 2: This is an image of text, please replace with text using the `windows-command-prompt`.
See examples from `_quick-start/step1-install-rsk-local-node.md` and `tutorials/ethereum-devs/setup-truffle-oz.md`.

Author: I would like to keep this content as an image, because of the colors. It is not a windows terminal, it is a truffle terminal.

truffle(develop)> simpleStorage.set(10)
{ tx:
   '0x0cf3d1f80f12dbfca47fd69af7f4671194c578eb5c60e905334c8765178d9bdf',
  receipt:
   { transactionHash:
      '0x0cf3d1f80f12dbfca47fd69af7f4671194c578eb5c60e905334c8765178d9bdf',
     transactionIndex: 0,
     blockHash:
      '0x604ab07250c373d74cd6c3665efa1935333e0cc97c0fa0a7d5898fbdcd7de693',
     blockNumber: 10,
     from: '0x1056f747cf4bc7710e178b2aeed4eb8c8506c728',
     to: '0x73f2aa5d251dbded6c950257124ea93bb00c0ec0',
     gasUsed: 41424,
     cumulativeGasUsed: 41424,
     contractAddress: null,
     logs: [],
     status: true,
     logsBloom:
      '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
     rawLogs: [] },
  logs: [] }
truffle(develop)> 
```

> The transaction generates a receipt. This is the answer the blockchain generates to a transaction.

Reviewer: ```suggestion
The transaction generates a receipt. This is the data the blockchain generates to a transaction.
```

Reviewer 2: ```suggestion
This is a transaction receipt, generated by the blockchain nodes, in response to the transaction request.
```

- RSK Mainnet dpath: `m/44’/137’/0’/0`
- RSK Testnet dpath: `m/44’/37310’/0’/0`

For more information, check [RSKIP57](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP57.md).

## Get the current gas price

**Gas** is the internal pricing for running a transaction or contract. 
When you send tokens, interact with a contract, send RBTC, or do anything else on the blockchain, you must pay for that computation. 
That payment is calculated as gas. 
In RSK, this is paid in **R-BTC**.
The **minimumGasPrice** is written in the block header by miners and establishes the minimum gas price that a transaction should have in order to be included in that block.

To get the **minimumGasPrice** do the following steps:

1. Run this query using cURL:

**Mainnet**

```shell
curl https://public-node.rsk.co/ \
    -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest",false],"id":1}'
```

**Testnet**

```shell
curl https://public-node.testnet.rsk.co/ \
    -X POST -H "Content-Type: application/json" \
    --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest",false],"id":1}'
```

> ![minimumGasPrice](/assets/img/tutorials/rsk-starter-box/image-08.png)

Reviewer 2: This is an image of text, please replace with text using the `windows-command-prompt`.
See examples from `_quick-start/step1-install-rsk-local-node.md` and `tutorials/ethereum-devs/setup-truffle-oz.md`.

Author: I would like to keep this content as an image. It is not a windows terminal, it is a `git bash` terminal.

2. Find in the result the field **_minimumGasPrice_**

For more information about the **Gas** and **minimumGasPrice** please visit the [gas](/rsk/rbtc/gas/ "Gas - RSK Developers Portal") page.

3. In `truffle-config.js`, check the value of `gasPrice` configuration and update it with the value found in the previous step.

![gasPrice](/assets/img/tutorials/rsk-starter-box/image-09.png)

> # Truffle Console connect to RSK network

Reviewer: ```suggestion
# Using Truffle Console to connect to the RSK network
```

 Run the development console for any RSK network.

 **Mainnet**

```shell
truffle console --network mainnet
```

**Testnet**

```shell
truffle console --network testnet
```

> This action instructs Truffle to connect to an RSK public node and grants it permission to control the accounts created with your mnemonic through HD wallet provider.

Reviewer: ```suggestion
This action instructs Truffle to connect to an RSK public node and grants it permission to control the accounts created with your mnemonic through the **HD wallet provider**.
```

> I will do it on Testnet:

Reviewer: ```suggestion
I will connect to the Testnet network:
```

```windows-command-prompt
C:\RSK\rsk-starter>truffle console --network testnet
truffle(testnet)>  
```

## Get your accounts

In the Truffle console, enter:

```javascript
const accounts = await web3.eth.getAccounts()
```

```windows-command-prompt
truffle(testnet)>  const accounts = await web3.eth.getAccounts()
undefined
truffle(testnet)>
```

To list the accounts:

```javascript
accounts
```

```windows-command-prompt
truffle(testnet)> accounts
[ '0xCd70794c2F3C657310eF13b6FF3Ec2d112513B39',
  '0x2325b62A11F07BB48fc7bD5B1F472Be7045d2f2d',
  '0x3fBF9136bDa714FD95c1e75DB5dd3b8d59fa1A58',
  '0xd7A88F54230e18159C528B682ba61273E4e1009e',
  '0xb7C8e2aD92fb57a0cB8fcF7bbA11e81E4e664C2E' ]
truffle(testnet)>
```

To view each account:

```javascript
accounts[0]
accounts[1]
```

> ![accounts](/assets/img/tutorials/rsk-starter-box/image-10.png)

Reviewer 2: This is an image of text, please replace with text using the `windows-command-prompt`.
See examples from `_quick-start/step1-install-rsk-local-node.md` and `tutorials/ethereum-devs/setup-truffle-oz.md`.

Author: I would like to keep this content as an image, because of the colors. It is not a windows terminal, it is a truffle terminal.

In my example, the account[0] is `0xCd70794c2F3C657310eF13b6FF3Ec2d112513B39`. 

## Check balance

To check the balance of account[0], for example, run this command in Truffle console:

```javascript
(await web3.eth.getBalance(accounts[0])).toString()
```

```windows-command-prompt
truffle(testnet)> (await web3.eth.getBalance(accounts[0])).toString()
'0'
truffle(testnet)>  
```

The balance is 0 and we need some tR-BTC to pay gas fees,
which will be used to publish smart contracts and interact with them.

> We shall obtain some in the next step.

Reviewer: ```suggestion
We shall obtain some tR-BTC in the next step.
```

## Exit Truffle console

In the Truffle console, enter this command to exit the terminal:

```shell
.exit
```

```windows-command-prompt
truffle(testnet)> .exit

C:\RSK\rsk-starter>
```

# Get R-BTC

The Smart Bitcoin (R-BTC) is the token used to pay for the execution of transactions in RSK.

**Mainnet**

For the RSK Mainnet, get R-BTC from [an exchange](https://www.rsk.co/#exchanges-rsk).

**Testnet**

For the RSK Testnet, get tR-BTC from [our faucet](https://faucet.testnet.rsk.co/).

![faucet.testnet.rsk.co](/assets/img/tutorials/rsk-starter-box/image-12.png)

Enter your wallet address and pass the CAPTCHA.

Wait a few seconds…

![Received some R-BTCs](/assets/img/tutorials/rsk-starter-box/image-13.png)

You can see the transaction hash:
[0xe7a25985f019482d362a3be908f1c0b3dee612fcc78716b6a341d8ad6138ea95](https://explorer.testnet.rsk.co/tx/0xe7a25985f019482d362a3be908f1c0b3dee612fcc78716b6a341d8ad6138ea95)

## Recheck balance

To check balance again, run this command in the Truffle console:

```javascript
(await web3.eth.getBalance(accounts[0])).toString()
```

For my example on RSK Testnet using account `0xCd70794c2F3C657310eF13b6FF3Ec2d112513B39`:

> ![getBalance](/assets/img/tutorials/rsk-starter-box/image-11.png)

Reviewer 2: This is an image of text, please replace with text using the `windows-command-prompt`.
See examples from `_quick-start/step1-install-rsk-local-node.md` and `tutorials/ethereum-devs/setup-truffle-oz.md`.

Author: I would like to keep this content as an image, because of the colors. It is not a windows terminal, it is a truffle terminal.

Great! Now I have 50000000000000000, which means that I have 0.05 tR-BTC with 18 decimal places of precision.

# Deploy the smart contract on RSK network

> Let's now switch to interacting with a "real" blockchain,

Reviewer: ```suggestion
Let's switch to interacting with a "real" blockchain (Mainnet),
```

Author: No, I am talking about go from Truffle develop console to RSK network, but can be Mainnet or Testnet.

   > total cost:          0.000020249299725056 ETH

   > Saving migration to chain.
   > Saving artifacts
   -------------------------------------
   > Total cost:     0.000020249299725056 ETH

2_deploy_contracts.js
=====================

   Deploying 'SimpleStorage'
   -------------------------
   > transaction hash:    0x6d29215af9e0b12814008bfb0965de27b81f25246523de7bc18ee18462f16629
   > Blocks: 2            Seconds: 77
   > contract address:    0xD0d125a883755A9B9AC55628BDd6b434e7865002
   > block number:        907727
   > block timestamp:     1591378215
   > account:             0xCd70794c2F3C657310eF13b6FF3Ec2d112513B39
   > balance:             0.049968280798422272
   > gas used:            107369 (0x1a369)
   > gas price:           0.076777216 gwei
   > value sent:          0 ETH
   > total cost:          0.000008243492904704 ETH

   > Saving migration to chain.
   > Saving artifacts
   -------------------------------------
   > Total cost:     0.000008243492904704 ETH

Summary
=======
> Total deployments:   2
> Final cost:          0.00002849279262976 ETH

C:\RSK\rsk-starter>
```

Congratulations!

`Simple storage` is now published on the RSK network.

> ![Simple storage published](/assets/img/tutorials/rsk-starter-box/image-14.png)

Reviewer 2: This is an image of text, please replace with text using the `windows-command-prompt`.
See examples from `_quick-start/step1-install-rsk-local-node.md` and `tutorials/ethereum-devs/setup-truffle-oz.md`.

Author: This is the only that I have the 2 versions:
`windows-command-prompt` and the image.
Ok, I chose the `windows-command-prompt` version and deleted the image

> Make sure you have enough R-BTC to deploy it.
    
# Interact with a smart contract on RSK network

Interact with the simple storage smart contract using Truffle console connected to an RSK network.

> It's the same as we did for Truffle development console, but now it will be for a real blockchain!

Reviewer: ```suggestion
The same method applied when connecting to the Truffle development console, is the same method to be used, but this time we will be connecting to Mainnet!
```

Author: No, same the previous, I am talking about go from Truffle develop console to RSK network, but can be Mainnet or Testnet.

> Make sure you deploy the smart contract before executing this part.

The next commands will run inside the Truffle console.

> ## Connect with our contract published

Reviewer: ```suggestion
## Connect with our published contract
```

{ tx:
   '0xaea9d3d21e95045c56843925bf9ea84428ef623246ec6ff95276d2e4d0a5bd93',
  receipt:
   { transactionHash:
      '0xaea9d3d21e95045c56843925bf9ea84428ef623246ec6ff95276d2e4d0a5bd93',
     transactionIndex: 0,
     blockHash:
      '0x3df23ea25148a4522831b453a08ee254e6f39408918b524eb87df88c0a4e01b2',
     blockNumber: 907750,
     cumulativeGasUsed: 41684,
     gasUsed: 41684,
     contractAddress: null,
     logs: [],
     from: '0xcd70794c2f3c657310ef13b6ff3ec2d112513b39',
     to: '0xd0d125a883755a9b9ac55628bdd6b434e7865002',
     root: '0x01',
     status: true,
     logsBloom:
      '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
     rawLogs: [] },
  logs: [] }
truffle(testnet)>
```

> The transaction generates a receipt. This is the answer the blockchain generates to a transaction.

Reviewer: ```suggestion
The transaction generates a receipt. This is the data the blockchain generates to a transaction.
```

Author: It is not only data, it is the return, the response. What is the better word to use here?

Reviewer 2: ```suggestion
This is a transaction receipt, generated by the blockchain nodes, in response to the transaction request.
```

```javascript
(await web3.eth.getBalance(accounts[0])).toString()
```

```windows-command-prompt
truffle(testnet)> (await web3.eth.getBalance(accounts[0])).toString()
'0'
truffle(testnet)>  
```

The balance is 0 and we need some tR-BTC to pay gas fees,
which will be used to publish smart contracts and interact with them.
We shall obtain some tR-BTC in the next step.

## Exit Truffle console

In the Truffle console, enter this command to exit the terminal:

```shell
.exit
```

```windows-command-prompt
truffle(testnet)> .exit

C:\RSK\rsk-starter>
```

# Get R-BTC

The Smart Bitcoin (R-BTC) is the token used to pay for the execution of transactions in RSK.

**Mainnet**

For the RSK Mainnet, get R-BTC from [an exchange](https://www.rsk.co/#exchanges-rsk).

**Testnet**

For the RSK Testnet, get tR-BTC from [our faucet](https://faucet.testnet.rsk.co/).

![faucet.testnet.rsk.co](/assets/img/tutorials/rsk-starter-box/image-12.png)

Enter your wallet address and pass the CAPTCHA.

Wait a few seconds…

![Received some R-BTCs](/assets/img/tutorials/rsk-starter-box/image-13.png)

You can see the transaction hash:

> [0xe7a25985f019482d362a3be908f1c0b3dee612fcc78716b6a341d8ad6138ea95](https://explorer.testnet.rsk.co/tx/0xe7a25985f019482d362a3be908f1c0b3dee612fcc78716b6a341d8ad6138ea95)

Reviewer 2: ```suggestion
[`0xe7a25985f019482d362a3be908f1c0b3dee612fcc78716b6a341d8ad6138ea95`](https://explorer.testnet.rsk.co/tx/0xe7a25985f019482d362a3be908f1c0b3dee612fcc78716b6a341d8ad6138ea95)
```
